clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java

```diff
@@ -0,0 +1,199 @@
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static java.time.Duration.ofSeconds;
+import static org.ehcache.testing.StandardCluster.clusterPath;
+import static org.ehcache.testing.StandardCluster.leaseLength;
+import static org.ehcache.testing.StandardCluster.newCluster;
+import static org.ehcache.testing.StandardCluster.offheapResource;
+import static org.ehcache.testing.StandardTimeouts.eventually;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.fail;
+import static org.terracotta.utilities.test.rules.TestRetryer.OutputIs.CLASS_RULE;
+import static org.terracotta.utilities.test.rules.TestRetryer.tryValues;
+
+public class MultipleCacheReconnectTest {
+  private static TCPProxyManager proxyManager;
+  private static PersistentCacheManager cacheManager;
+
+  private static CacheConfiguration<Long, String> config = CacheConfigurationBuilder.newCacheConfigurationBuilder(
+    Long.class, String.class, ResourcePoolsBuilder.newResourcePoolsBuilder()
+      .with(ClusteredResourcePoolBuilder.clusteredDedicated("primary-server-resource", 1, MemoryUnit.MB)
+      )).withResilienceStrategy(new ThrowingResiliencyStrategy<>()).build();
+
+  @ClassRule
+  @Rule
+  public static final TestRetryer<Duration, Cluster> CLUSTER = tryValues(ofSeconds(1), ofSeconds(10), ofSeconds(30)).map(leaseLength -> newCluster().in(clusterPath()).withServiceFragment(offheapResource("primary-server-resource", 64) + leaseLength(leaseLength)).build()).outputIs(CLASS_RULE);
+
+  @BeforeClass
+  public static void initializeCacheManager() throws Exception {
+    proxyManager = TCPProxyManager.create(CLUSTER.get().getConnectionURI());
+    URI connectionURI = proxyManager.getURI();
+
+    CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder = CacheManagerBuilder.newCacheManagerBuilder().with(ClusteringServiceConfigurationBuilder.cluster(connectionURI.resolve("/crud-cm")).autoCreate(server -> server.defaultServerResource("primary-server-resource")));
+    cacheManager = clusteredCacheManagerBuilder.build(false);
+    cacheManager.init();
+  }
+
+  @AfterClass
+  public static void stopProxies() {
+    proxyManager.close();
+  }
+
+  @Test
+  public void cacheOpsDuringReconnection() throws Exception {
+
+    try {
+      Cache<Long, String> cache1 = cacheManager.createCache("clustered-cache1", config);
+      Cache<Long, String> cache2 = cacheManager.createCache("clustered-cache2", config);
+
+      expireLease();
+
+      CompletableFuture<Void> putFuture1 = CompletableFuture.runAsync(() -> {
+        cache1.put(0L, "firstcache");
+      });
+
+      CompletableFuture<Void> putFuture2 = CompletableFuture.runAsync(() -> {
+        cache2.put(0L, "secondcache");
+      });
+
+      try {
```

Comment:
This form would be preferable to the try/catch block:

```java
      assertThat(() -> putFuture1.get(5000, TimeUnit.MILLISECONDS),
        threw(Matchers.<Throwable>both(instanceOf(ExecutionException.class))
          .and(hasCause(hasCause(hasCause(hasCause(instanceOf(ReconnectInProgressException.class))))))));
```

@chrisdennis recommend a different mechanism than the chain of `hasCause` methods.